wurst/math/Quaternions.wurst

```diff
@@ -0,0 +1,107 @@
+    return this.x*q.x + this.y*q.y + this.z*q.z + this.w*q.w
+
+/** Represents the composition of two rotations. */
+public function quat.cross(quat q) returns quat
+    return quat(
+        this.x*q.w + this.y*q.z - this.z*q.y + this.w*q.x,
+        -this.x*q.z + this.y*q.w + this.z*q.x + this.w*q.y,
+        this.x*q.y - this.y*q.x + this.z*q.w + this.w*q.z,
+        -this.x*q.x - this.y*q.y - this.z*q.z + this.w*q.w)
+
+/** Represents the same rotation in opposite direction. */
+public function quat.conj() returns quat
+    return quat(-this.x, -this.y, -this.z, this.w)
+
+public function quat.length() returns real
+    return SquareRoot(this.x.squared() + this.y.squared() + this.z.squared() + this.w.squared())
+
+/** Returns zero quaternion if lenght is zero. */
+public function quat.norm() returns quat
+    var length = this.length()
+    return length > 0.002 ? quat(this.x/length, this.y/length, this.y/length, this.w/length) : ZEROQ
+
+public function quat.toMatrix() returns mat3
+    var q = this.length().isBetween(0.998, 1.002) ? this : this.norm()
+    var x = q.x
+    var y = q.y
+    var z = q.z
+    var w = q.w
+    return mat3(
+        1 - 2*y*y - 2*z*z,2*x*y - 2*z*w, 2*x*z + 2*y*w,
+        2*x*y + 2*z*w, 1 - 2*x*x - 2*z*z, 2*y*z - 2*x*w,
+        2*x*z - 2*y*w, 2*y*z + 2*x*w, 1 - 2*x*x - 2*y*y)
+
+/** Spherical Linear intERPolation. Interpolates between two rotations.
+q - targeted quaternion. p - progress from 0 to 1.*/
+public function quat.slerp(quat q, real p) returns quat
+    var dot = this.dot(q)
+    if dot.abs() < 0.002
+        return q
+    var theta = dot.acos()
+    return ((1 - p)*theta).sin()/theta.sin()*this + (p*theta).sin()/theta.sin()*q
+
+/** Spherical Linear intERPolation. Interpolates between two rotations.
+Always choosing the SHORTEST way between rotations.
+q - targeted quaternion. p - progress from 0 to 1.*/
+public function quat.slerpShort(quat q, real p) returns quat
+    var dot = this.dot(q)
+    if dot.abs() < 0.002
+        return q
+    // choosing the shortest way
+    return dot < 0 ? this.slerp(quat(-q.x, -q.y, -q.z, -q.w), p) : this.slerp(q, p)
+
+/** Spherical Linear intERPolation. Interpolates between two rotations.
+Always choosing the LONGEST way between rotations.
+q - targeted quaternion. p - progress from 0 to 1.*/
+public function quat.slerpLong(quat q, real p) returns quat
+    var dot = this.dot(q)
+    if dot.abs() < 0.002
+        return q
+    // choosing the longest way
+    return dot < 0 ? this.slerp(q, p) : this.slerp(quat(-q.x, -q.y, -q.z, -q.w), p)
+    
+public function quat.toString() returns string
+    return "Quaternion [X {0}, Y {1}, Z {2}, W {3}]".format(this.x.toString(), this.y.toString(), this.z.toString(), this.w.toString())
+
+/** Extracts Euler-angles (Tait-Bryan) from a quaternion. Order of result angles is X-Y-Z.
+There's no unic way for quaternions thus we have to use quat-matrix conversion first. */
+public function quat.toEuler() returns vec3

```

Comment:
Using vec3 for representing a euler angle is a bit misleading here. While technically the representation is the same, semantically these are very different mathematical concepts, and most operations applied to vectors do not make sense in the context of euler angles.

Unless what you mean to imply here by Euler angle is actually a directional unit-vector... Which afaik is not an Euler angle at all.

A separate tuple for euler angles (and maybe even a separate package with functions to operate on them) would be ideal IMO.

Reply:
I just realized you mentioned Tait-Bryan angles in there.
Is there a particular reason for using those instead of classical Euler angles? As far as I know, the typical usage of Euler angles in games is the classical Euler angle, i.e. Pitch, Yaw, Roll, and not Yaw, Pitch, Roll